src/QsCompiler/CompilationManager/CompilationUnit.cs

```diff
@@ -686,5 +701,71 @@ internal LocalDeclarations TryGetLocalDeclarations(FileContentManager file, Posi
             var declarations = implementation?.LocalDeclarationsAt(pos.Subtract(specPos), includeDeclaredAtPosition);
             return this.PositionedDeclarations(parentCallable, callablePos, specPos, declarations); 
         }
+
+        /// <summary>
+        /// Returns a short tag name for the given source file. The tag is not unique. If multiple source files are
+        /// being tagged, make sure there are no name conflicts.
+        /// </summary>
+        /// <param name="source">The source file to get the tag for.</param>
+        /// <returns>A short tag name for the given source file.</returns>
+        internal static string GetTagForSource(string source) =>
+            Regex.Replace(Path.GetFileNameWithoutExtension(source), "[^A-Za-z0-9_]", "");
+
+        /// <summary>
+        /// Embeds the tag name into the qualified name.
+        /// </summary>
+        /// <param name="tag">The tag name to embed.</param>
+        /// <param name="name">The qualified name in which to embed the tag name.</param>
+        /// <returns>A new qualified name with the tag embedded into it.</returns>
+        internal static QsQualifiedName EmbedTag(string tag, QsQualifiedName name) =>
+            // TODO: We might need to change the format to make it easier to reverse later.
```

Comment:
Regarding easily reversible, e.g. the pattern used in UniqueVariableNames has a tested regex pattern.